pkg/webhooks/pytorch/pytorchjob_webhook.go

```diff
@@ -0,0 +1,146 @@
+	job := newObj.(*trainingoperator.PyTorchJob)
+	log := ctrl.LoggerFrom(ctx).WithName("pytorchjob-webhook")
+	log.V(5).Info("Validating update", "pytorchJob", klog.KObj(job))
+	warnings, errs := validatePyTorchJob(job)
+	return warnings, errs.ToAggregate()
+}
+
+func (w *Webhook) ValidateDelete(context.Context, runtime.Object) (admission.Warnings, error) {
+	return nil, nil
+}
+
+func validatePyTorchJob(job *trainingoperator.PyTorchJob) (admission.Warnings, field.ErrorList) {
+	var allErrs field.ErrorList
+	var warnings admission.Warnings
+
+	if errors := apimachineryvalidation.NameIsDNS1035Label(job.ObjectMeta.Name, false); len(errors) != 0 {
+		allErrs = append(allErrs, field.Invalid(field.NewPath("metadata").Child("name"), job.Name, fmt.Sprintf("should match: %v", strings.Join(errors, ","))))
+	}
+	ws, err := validateSpec(job.Spec)
+	warnings = append(warnings, ws...)
+	allErrs = append(allErrs, err...)
+	return warnings, allErrs
+}
+
+func validateSpec(spec trainingoperator.PyTorchJobSpec) (admission.Warnings, field.ErrorList) {
+	var allErrs field.ErrorList
+	var warnings admission.Warnings
+
+	if spec.NprocPerNode != nil && spec.ElasticPolicy != nil && spec.ElasticPolicy.NProcPerNode != nil {
+		elasticNProcPerNodePath := specPath.Child("elasticPolicy").Child("nProcPerNode")
+		nprocPerNodePath := specPath.Child("nprocPerNode")
+		allErrs = append(allErrs, field.Forbidden(elasticNProcPerNodePath, fmt.Sprintf("must not be used with %s", nprocPerNodePath)))
+		warnings = append(warnings, fmt.Sprintf("%s is deprecated, use %s instead", elasticNProcPerNodePath.String(), nprocPerNodePath.String()))
+	}
+	allErrs = append(allErrs, validatePyTorchReplicaSpecs(spec.PyTorchReplicaSpecs)...)
+	return warnings, allErrs
+}
+
+func validatePyTorchReplicaSpecs(rSpecs map[trainingoperator.ReplicaType]*trainingoperator.ReplicaSpec) field.ErrorList {
+	var allErrs field.ErrorList
+
+	if rSpecs == nil {
+		allErrs = append(allErrs, field.Required(pytorchReplicaSpecPath, "must be required"))
+	}
+	for rType, rSpec := range rSpecs {
+		rolePath := pytorchReplicaSpecPath.Key(string(rType))
+		containersPath := rolePath.Child("template").Child("spec").Child("containers")
+
+		// Make sure the replica type is valid.
+		validRoleTypes := []string{
+			string(trainingoperator.PyTorchJobReplicaTypeMaster),
+			string(trainingoperator.PyTorchJobReplicaTypeWorker),
+		}
+		if !slices.Contains(validRoleTypes, string(rType)) {
+			allErrs = append(allErrs, field.NotSupported(rolePath, rType, validRoleTypes))
+		}
+
+		if rSpec == nil || len(rSpec.Template.Spec.Containers) == 0 {
+			allErrs = append(allErrs, field.Required(containersPath, "must be specified"))
+		}
+
+		// Make sure the image is defined in the container
+		defaultContainerPresent := false
+		for idx, container := range rSpec.Template.Spec.Containers {
+			if container.Image == "" {
+				allErrs = append(allErrs, field.Required(containersPath.Index(idx).Child("image"), "must be required"))
```

Comment:
We can't somehow invoke the Pod validation here to rely on default Kubernetes Pod validation, right ?
Thus, we should verify the required Pod container parameters (e.g. image) by ourselves ?

Reply:
Yes, there is no way to use any library since these validations are not exported: https://github.com/kubernetes/kubernetes/blob/9791f0d1f39f3f1e0796add7833c1059325d5098/pkg/apis/core/validation/validation.go

Reply:
Yes, there is no way to use any library since these validations are not exported: https://github.com/kubernetes/kubernetes/blob/9791f0d1f39f3f1e0796add7833c1059325d5098/pkg/apis/core/validation/validation.go